shared/src/Datadog.Trace.ClrProfiler.Native/crashhandler.cpp

```diff
@@ -0,0 +1,409 @@
+
+        if (_context.Environ != nullptr)
+        {
+            delete[] _context.Environ;
+        }
+
+        _context.Environ = nullptr;
+        _context.EnvironLength = 0;
+    }
+
+    CrashHandler::CrashHandler()
+        : _context(),
+        _crashHandler()
+    {
+    }
+
+    bool CrashHandler::Register()
+    {
+        if (!_crashHandler.empty())
+        {
+            Log::Warn("Crashtracking - Trying to initialize the crash handler twice");
+            return false;
+        }
+
+        auto dllPath = GetCurrentDllPath();
+
+        if (dllPath.empty())
+        {
+            Log::Warn("Crashtracking - Could not register the crash handler: error when retrieving the path of the DLL");
+            return false;
+        }
+
+        GetDatadogEnvironmentBlock(_context.Environ, _context.EnvironLength);
+
+        // Register the crash handler in the registry
+        // Windows expects a DWORD value with the full path of the DLL as the name, 
+        // in SOFTWARE\Microsoft\Windows\Windows Error Reporting\RuntimeExceptionHelperModules.
+        // The key can be located either in HKLM or HKCU. The MSI will create the value in HKLM,
+        // but if it's missing we add it to HKCU.
+        HKEY hKey;
+        LPCWSTR subKey = L"SOFTWARE\\Microsoft\\Windows\\Windows Error Reporting\\RuntimeExceptionHelperModules";
+        DWORD value = 1;
+
+        if (!RegistryValueExists(HKEY_LOCAL_MACHINE, subKey, dllPath.c_str()))
+        {
+            // Open the key
+            DWORD disposition;
+            auto result = RegCreateKeyEx(HKEY_CURRENT_USER, subKey, 0, NULL, 0, KEY_SET_VALUE, NULL, &hKey, &disposition);
+
+            if (result != ERROR_SUCCESS)
+            {
+                // Failing to set the registry is not a fatal error: in the worst case scenario the crash handler just won't be invoked by WER
+                Log::Warn("Crashtracking - Failed to create registry key: ", GetLastError());
+            }
+            else
+            {
+                // Set the value
+                result = RegSetValueEx(hKey, dllPath.c_str(), 0, REG_DWORD, reinterpret_cast<const BYTE*>(&value), sizeof(value));
+
+                if (result != ERROR_SUCCESS)
+                {
+                    Log::Warn("Crashtracking - Failed to set registry value: ", GetLastError());
+                }
+
+                RegCloseKey(hKey);
+            }
+        }
+
+        // The profiler API is not initialized yet, so we look for clr.dll/coreclr.dll
+        // to know if we're running with .NET Framework or .NET Core
+        bool isDotnetCore = true;
+        HMODULE hModule = GetModuleHandle(L"coreclr.dll");
+
+        if (hModule == NULL)
+        {
+            hModule = GetModuleHandle(L"clr.dll");
+            isDotnetCore = false;
+
+            if (hModule == NULL)
+            {
+                Log::Warn("Crashtracking - Failed to get module handle for coreclr.dll or clr.dll");
+                return false;
```

Comment:
I assume we don't need to explicitly delete environ in this error case, because you're handling it in the destructor now, right?